tests/unit/test_charm_2.py

```diff
@@ -1,58 +1,166 @@
+import re
+import json
 import unittest
+import collections
 import unittest.mock
+
 import ops.pebble
 import ops.testing
+import mysql.connector
 
 from charm import WordpressCharm
 from exceptions import *
 
 
-class TestWordpressK8s(unittest.TestCase):
-    def setUp(self):
-        container_fs = {}
-        self.container_fs = container_fs
-        installed_database_hosts = set()
-        self.installed_database_hosts = installed_database_hosts
+class WordpressMock:
+    """WordPress wp-cli command run and database simulation system for unit tests"""
+    def __init__(self):
+        self._database = {}
+        self._container_fs = {}
+        self._database_credentials = {}
+        self._themes = set(WordpressCharm._WORDPRESS_DEFAULT_THEMES)
+        self._plugins = set(WordpressCharm._WORDPRESS_DEFAULT_PLUGINS)
+        self._patch = None
+
+    def _get_current_database_config(self):
+        wp_config = self._container_fs.get(WordpressCharm._WP_CONFIG_PATH)
+        if wp_config is None:
+            return None
+        db_info = {}
+        for db_key in ['db_host', 'db_name', 'db_user', 'db_password']:
+            db_value = re.findall(f"define\\( '{db_key.upper()}', '([^']+)' \\);", wp_config)
+            if not db_value:
+                raise ValueError(f"{db_key} is missing in wp-config.php")
+            if len(db_value) > 1:
+                raise ValueError(f"multiple {db_key} definitions")
+            db_info[db_key] = db_value[0]
+        return db_info
+
+    def _mock_database_connect(self, db_info):
+        credential_key = db_info["db_host"], db_info["db_name"]
+        if credential_key not in self._database_credentials:
+            raise mysql.connector.Error(
+                msg=f"Can't connect to MySQL server on '{db_info['db_host']}:3306' (2003)",
+                errno=2003
+            )
+        for credential in self._database_credentials[credential_key]:
+            if (
+                    credential["db_user"] == db_info["db_user"] and
+                    credential["db_password"] == db_info["db_password"]
+            ):
+                return
+        raise mysql.connector.Error(
+            msg=f"Access denied for user '{db_info['db_user']}'@* (using password: *)",
+            errno=1045
+        )
 
+    def _simulate_run_wp_cli(self, cmd):
+        Result = collections.namedtuple("WordpressCliExecResult", "return_code stdout stderr")
```

Comment:
I remember this being defined elsewhere before, is that the case? If so, would it be worth sharing this definition?

Reply:
`_ ExecResult` is different in terms of fields.
```python
    _ExecResult = collections.namedtuple("ExecResult", "success result message")
```